File: app/src/main/java/edu/sunypoly/inventoryapp/InventoryItem.java
package edu.sunypoly.inventoryapp;

import android.os.Bundle;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectInputStream;
import java.io.ObjectOutput;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.HashMap;

public class InventoryItem implements Serializable {
    private int id;
    private int barcode;
    private String qr;
    private String name;
    private String type;
    private int serial;
    private String room;
    private String brand;
    private String acquired;

    public static final String ID = "id";
    public static final String BARCODE = "barcode";
    public static final String QR = "qr";
    public static final String NAME = "name";
    public static final String TYPE = "type";
    public static final String SERIAL = "serial";
    public static final String ROOM = "room";
    public static final String BRAND = "brand";
    public static final String ACQUIRED = "acquired";

    public InventoryItem() {
    }

    public InventoryItem(final int id, final int barcode, final String qr, final String name,
                         final String type, final int serial, final String room, final String brand,
                         final String acquired) {
        this.id = id;
        this.barcode = barcode;
        this.qr = qr;
        this.name = name;
        this.type = type;
        this.serial = serial;
        this.room = room;
        this.brand = brand;
        this.acquired = acquired;

    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
//        fieldsMap.put(ID, "" + id);
    }

    public int getBarcode() {
        return barcode;
    }

    public void setBarcode(int barcode) {
        this.barcode = barcode;
//        fieldsMap.put(BARCODE, "" + barcode);
    }

    public String getQr() {
        return qr;
    }

    public void setQr(String qr) {
        this.qr = qr;
//        fieldsMap.put(QR, qr);
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
//        fieldsMap.put(NAME, name);
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
//        fieldsMap.put(TYPE, type);
    }

    public int getSerial() {
        return serial;
    }

    public void setSerial(int serial) {
        this.serial = serial;
//        fieldsMap.put(SERIAL, "" + serial);
    }

    public String getRoom() {
        return room;
    }

    public void setRoom(String room) {
        this.room = room;
//        fieldsMap.put(ROOM, room);
    }

    public String getBrand() {
        return brand;
    }

    public void setBrand(String brand) {
        this.brand = brand;
//        fieldsMap.put(BRAND, brand);
    }

    public String getAcquired() {
        return acquired;
    }

    public void setAcquired(String acquired) {
        this.acquired = acquired;
//        fieldsMap.put(ACQUIRED, acquired);
    }

    @Override
    public String toString() {
        return "InventoryItem{" +
                "id=" + id +
                ", barcode=" + barcode +
                ", qr='" + qr + '\'' +
                ", name='" + name + '\'' +
                ", type='" + type + '\'' +
                ", serial=" + serial +
                ", room='" + room + '\'' +
                ", brand='" + brand + '\'' +
                ", acquired='" + acquired + '\'' +
                '}';
    }

    public Bundle toBundle() {
        Bundle bundle = new Bundle();

        bundle.putInt(ID, id);
        bundle.putInt(BARCODE, barcode);
        bundle.putString(QR, qr);
        bundle.putString(NAME, name);
        bundle.putString(TYPE, type);
        bundle.putInt(SERIAL, serial);
        bundle.putString(ROOM, room);
        bundle.putString(BRAND, brand);
        bundle.putString(ACQUIRED, acquired);

        return bundle;
    }

    HashMap<String, String> getFields() {
        return new HashMap<String, String>() {{
            put(ID, "" + id);
            put(BARCODE, "" + barcode);
            put(QR, qr);
            put(NAME, name);
            put(TYPE, type);
            put(SERIAL, "" + serial);
            put(ROOM, room);
            put(BRAND, brand);
            put(ACQUIRED, acquired);
        }};
    }

    byte[] toByteArray() throws IOException {
        try (ByteArrayOutputStream bos = new ByteArrayOutputStream();
             ObjectOutput out = new ObjectOutputStream(bos)) {
            out.writeObject(this);
            return bos.toByteArray();
        }
    }

    static InventoryItem fromByteArray(byte[] bytes) throws IOException, ClassNotFoundException {
        try (ByteArrayInputStream bis = new ByteArrayInputStream(bytes);
             ObjectInput in = new ObjectInputStream(bis)) {
            return (InventoryItem) in.readObject();
        }
    }

     boolean contains(String s){
        return (id + "" + barcode + qr + name + type + serial + room + brand + acquired).contains(s);
     }
}

File: app/src/main/java/edu/sunypoly/inventoryapp/InventoryItem.kt
package edu.sunypoly.inventoryapp

import android.os.Bundle

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.io.Serializable
import java.util.HashMap

/**
 * Stores inventory item data
 * Used by GSON to convert between JSONs and Objects.
 */
class InventoryItem(//        fieldsMap.put(ID, "" + id);
        var id: Int,//        fieldsMap.put(BARCODE, "" + barcode);
        var barcode: Int,//        fieldsMap.put(QR, qr);
        var qr: String,//        fieldsMap.put(NAME, name);
        var name: String,//        fieldsMap.put(TYPE, type);
        var type: String,//        fieldsMap.put(SERIAL, "" + serial);
        var serial: String,//        fieldsMap.put(ROOM, room);
        var room: String,//        fieldsMap.put(BRAND, brand);
        var brand: String,//        fieldsMap.put(ACQUIRED, acquired);
        var acquired: String) : Serializable {

    /**
     * Returns a HashMap of all the id keys
     * @return
     */
    internal val fields: HashMap<String, String>
        get() = object : HashMap<String, String>() {
            init {
                put(ID, "" + id)
                put(BARCODE, "" + barcode)
                put(QR, qr)
                put(NAME, name)
                put(TYPE, type)
                put(SERIAL, "" + serial)
                put(ROOM, room)
                put(BRAND, brand)
                put(ACQUIRED, acquired)
            }
        }

    override fun toString(): String {
        return "InventoryItem{" +
                "id=" + id +
                ", barcode=" + barcode +
                ", qr='" + qr + '\''.toString() +
                ", name='" + name + '\''.toString() +
                ", type='" + type + '\''.toString() +
                ", serial=" + serial +
                ", room='" + room + '\''.toString() +
                ", brand='" + brand + '\''.toString() +
                ", acquired='" + acquired + '\''.toString() +
                '}'.toString()
    }

    /**
     * Returns the item as a Bundle for passing it through Intents, etc
     * (Alternate method to the byte[] stuff below)
     * @return Bundle
     */
    fun toBundle(): Bundle {
        val bundle = Bundle()

        bundle.putInt(ID, id)
        bundle.putInt(BARCODE, barcode)
        bundle.putString(QR, qr)
        bundle.putString(NAME, name)
        bundle.putString(TYPE, type)
        bundle.putString(SERIAL, serial)
        bundle.putString(ROOM, room)
        bundle.putString(BRAND, brand)
        bundle.putString(ACQUIRED, acquired)

        return bundle
    }

    /**
     * Used to convert the object into a form that can be passed through an Intent
     * @return
     * @throws IOException
     */
    @Throws(IOException::class)
    internal fun toByteArray(): ByteArray {
        ByteArrayOutputStream().use { bos ->
            ObjectOutputStream(bos).use { out ->
                out.writeObject(this)
                return bos.toByteArray()
            }
        }
    }

    /**
     * Used by SearchActivity to nicely check if this item contains the specified query.
     * @param s
     * @return
     */
    internal operator fun contains(s: String): Boolean {
        return (id.toString() + "" + barcode + qr + name + type + serial + room + brand + acquired).contains(s)
    }

    companion object {

        const val ID = "id"
        const val BARCODE = "barcode"
        const val QR = "qr"
        const val NAME = "name"
        const val TYPE = "type"
        const val SERIAL = "serial"
        const val ROOM = "room"
        const val BRAND = "brand"
        const val ACQUIRED = "acquired"

        /**
         * Used to convert the item back from a byte[] after being received from an intent
         * @param bytes
         * @return
         * @throws IOException
         * @throws ClassNotFoundException
         */
        @Throws(IOException::class, ClassNotFoundException::class)
        internal fun fromByteArray(bytes: ByteArray): InventoryItem {
            ByteArrayInputStream(bytes).use { bis -> ObjectInputStream(bis).use { `in` -> return `in`.readObject() as InventoryItem } }
        }
    }
}
